core/src/jsMain/kotlin/Peripheral.kt

```diff
@@ -128,56 +125,78 @@ public class JsPeripheral internal constructor(
     private val gatt: BluetoothRemoteGATTServer
         get() = bluetoothDevice.gatt!! // fixme: !!
 
-    private var connectJob: Deferred<Unit>? = null
-
-    private fun connectAsync() = scope.async(start = LAZY) {
-        logger.info { message = "Connecting" }
-        _state.value = State.Connecting.Bluetooth
-
+    private val connectJob: SharedRepeatableTask = scope.sharedRepeatableTask {
         try {
-            registerDisconnectedListener()
-
-            gatt.connect().await() // todo: Catch appropriate exception to emit State.Rejected.
-            _state.value = State.Connecting.Services
-
-            discoverServices()
-            onServicesDiscovered(ServicesDiscoveredPeripheral(this@JsPeripheral))
-            _state.value = State.Connecting.Observes
-            logger.verbose { message = "Configuring characteristic observations" }
-            observers.onConnected()
+            openConnection()
         } catch (t: Throwable) {
             logger.error(t) { message = "Failed to connect" }
-            disconnectGatt()
+            disconnectJob.launch()
             throw t
         }
-
-        logger.info { message = "Connected" }
-        _state.value = State.Connected
     }
 
-    private fun closeConnection() {
-        observationListeners.clear()
-        disconnectGatt()
-        unregisterDisconnectedListener()
-        _state.value = State.Disconnected()
+    private val disconnectJob: SharedRepeatableTask = scope.sharedRepeatableTask {
+        connectJob.join()
```

Comment:
Should this be `cancelAndJoin`?

Reply:
Nope. If there is a connect in-flight it is polite to let it complete before commencing the disconnection - less chance of the underlying system getting all twisted up.

Reply:
Gotchya. :+1: